### PR TITLE
support zsh; BASH_SOURCE does not work zsh

### PR DESCRIPTION
**Pull request by hanmoi-choi**
I had issue with zsh because BASH_SOURCE env does not work with my zsh

### CODE REVIEW

**Comment by mrowe** [comment]
> I had an issue with zsh

This tool _is_ called **bash** my AWS...

**Comment by hanmoi-choi** [comment]
I am agree with @mrowe so asked Collin and he told me that this would support `Zsh` as well so I tried.

**Comment by hanmoi-choi** [comment]
Also curious about that why you like to limit this good script to only BASH. Why do you hate users using ZSH? :)

**Comment by mrowe** [comment]
Ok, well there are probably better ways to determine if you are running in `zsh` (since it may not be in `/bin/zsh`). E.g. the first couple of results at https://www.google.com.au/search?q=how+to+tell+if+you+are+running+zsh

**Comment by hanmoi-choi** [comment]
Thanks a lot I will update my PR based on your feedback

**Comment by mbailey** [comment]
You're the first zsh user to contribute a fix @hanmoi-choi . Thanks, will get back to you about it.

**Comment by mrowe** [comment]
Since you're investing all this effort in it... :-) seems like there might be a better way of finding the `shared.inc` file that isn't shell-dependent? Maybe @mbailey has some ideas?

**Comment by hanmoi-choi** [comment]
@mrowe I like your link :). maybe using `ZSH_VERSION` would be better I reckon.
And curious about `else` case. do I have to exit 0 with echo `use either bash or zsh` kinda?

**Comment by mrowe** [comment]
I'm suggesting maybe there's some other way to find the shared file that works in any shell, so you don't need all those conditionals at all.

Maybe something with `$0`? I don't know, I haven't considered all the use cases. But the less you depend on specific shell features the more useful the code will be.

**Comment by hanmoi-choi** [comment]
I agree with @mrowe not like to rely on specific Shell but unfortunately my Shell skill is just intermediate so don't know how to manage it.

Could you let me know?

**Comment by hanmoi-choi** [comment]
`source` function could handle relative path?

**Comment by mrowe** [comment]
The problem is relative to _what_, since you probably aren't invoking the command from the directory in which it's stored. So you have to somehow get the absolute path of the file or its parent.

I'll leave it to @mbailey I think, since he knows this code much better than I do. :)

**Comment by hanmoi-choi** [comment]
But the code is not directly called from command line only referred from the same dir. it seems like shared utility functions so using relative path would be fine I reckon.
the function's naming starts with `__` which would stand for private method?

**Comment by hanmoi-choi** [comment]
Anyway, I will wait for @mbailey feedback as well. Thanks

**Comment by jonhiggs** [comment]
`if [[ ${SHELL} = "/bin/zsh" ]]; then` could be something like `if [[ $(basename ${SHELL}) == "zsh" ]]; then`.

**Comment by hanmoi-choi** [comment]
@jonhiggs Thanks, it looks good and I have learned about `basename` as well.

**Comment by michaelneale** [comment]
BASH_SOURCE doesn't work on windows XP

File: lib/instance-functions [hunk 1]
```diff
@@ -3,8 +3,11 @@
 # instance-functions
 #
 # List, run, start, stop and ssh to Amazon AWS EC2 instances
-
-source $(dirname ${BASH_SOURCE[0]})/shared.inc
+if [[ ${SHELL} = "/bin/zsh" ]]; then
```
Comment (vorlock) [inline lib/instance-functions]: This is not going to work ok Debian and it's derivatives where zsh is in `/usr/bin/zsh`

File: lib/instance-functions [hunk 2]
```diff
@@ -3,8 +3,11 @@
...
+  source $(dirname ${(%):-%N})/shared.inc
```
Comment (vorlock) [inline lib/instance-functions]: Is it not better to have `source $(dirname ${(%):-%x})/shared.inc` ?

**Comment by hootymcboob** [comment]
I don't see how this can work on AmigaOS

**Comment by mbailey** [comment]
You're not the first person to ask for zsh support but you are the first to contribute a pull request. Contributions are currency in open source. 

bash-my-aws is written to run under Bash as clearly stated in its prerequisites section in the [README](https://github.com/search?q=bash_source&type=Code&utf8=%E2%9C%93)

There are a number of reasons I do not plan to claim bash-my-aws is 'zsh compliant' but the main ones are that Bash is everywhere and I don't know enough about zsh to want to make that commitment.

BASH_SOURCE is [pretty commonly used](https://github.com/search?q=bash_source&type=Code&utf8=%E2%9C%93) to get the location of the file it's found in. You solution may solve the problem of it's absence from zsh but are there other bash'isms that don't work in zsh? Are there other shells that would require further complexity to the codebase? I can't sign up to support something I have no experience with or interest in using.

Perhaps a better approach would be for you to find a way to run bma functions using Bash.

For example, you could use the wrapper I wrote last Hack Day for running in a separate process?

```bash
MBA% ~/.bash-my-aws/bin/bma regions
eu-west-1
ap-southeast-1
ap-southeast-2
eu-central-1
ap-northeast-2
ap-northeast-1
us-east-1
sa-east-1
us-west-1
us-west-2
```

This is a very simple workaround that might work for anyone not using Bash as their shell.

```bash
#!/bin/bash

# bma
#
# Provides alternative to loading the functions into your
# shell environment.
#
# Useful for calling a function from another process (like 'env')
#
# e.g. env FOO=BAR bma instances
#
# Add it to your PATH or copy/link it to somewhere that is
#
# Assumes you have installed bash-my-aws to standard location

for f in ~/.bash-my-aws/lib/*-functions; do source $f; done

$@
```
Interested to hear if you could solve your issue by wrapping it rather than if-fecting the main codebase.

**Pull request closed without merging**